### ISSUE TITLE
Extend sharded composite pattern to gather_logprobs

### ISSUE THREAD

**Issue by alinakhanTT** [tenstorrent/tt-xla#5137]
## Summary

Extend the sharded composite sampling pattern introduced in [tt-xla#5004](https://github.com/tenstorrent/tt-xla/pull/5004) to the `gather_logprobs` path. After #5004, sampling no longer all_gathers the full vocab on 1D and 2D mesh TP, but `gather_logprobs` still does so whenever the user requests logprobs. Migrating it to the same local op plus small collective pattern is expected to yield roughly a 1000x reduction in per step wire bandwidth on the logprobs requested path.

## Background

Per [tt-xla#5004](https://github.com/tenstorrent/tt-xla/pull/5004), logits leave `compute_logits` vocab sharded as `[batch, vocab/N]` and the sampling path consumes them sharded via `tenstorrent.topk`. `gather_logprobs` undoes that on the logprobs requested path: at entry it applies `sharding_constraint_tensor(logits, mesh, (None, None))`, forcing an `all_gather` of `[batch, vocab]` so that `logsoftmax`, `topk`, `gather`, and `count_ge` can run replicated. At `batch=32`, `vocab=128k`, `fp32`, that is roughly 16 MB across the model axis per call.

The path is only exercised when the caller requests logprobs, but for workloads that do (eval, RLHF, structured generation with logprob scoring) it is the dominant per step cost on sharded TP.

### 1. `logsoftmax` via shard local `logsumexp` plus `all_reduce`

```
local_lse  = logsumexp(local_logits, dim=-1)        # [batch]
global_lse = log_sum_exp_all_reduce(local_lse)      # [batch], collective
local_logprobs = local_logits - global_lse[:, None] # [batch, vocab/N]
```

Wire bytes per shard, per step: `batch * 4`.

Requires a new `tenstorrent.logsumexp` composite plus custom sharding rule in tt-mlir. Same shape as the `tenstorrent.topk` rule, but the inter shard combine is log_sum_exp rather than merge with shard offsets.

### 2. `topk(num_logprobs)` via existing `tenstorrent.topk`

Already implemented in [tt-mlir#8601](https://github.com/tenstorrent/tt-mlir/pull/8601). No new compiler work, only a call site change in `sampler.gather_logprobs` to route through `composite_topk` when `vocab_sharded=True`.

Wire bytes per shard, per step: `batch * num_logprobs * 4`.

### 3. Sharded index gather for `token_logprobs`

`token_ids` carry global vocab positions and `logprobs` is vocab sharded. Pattern:

```
in_shard  = (token_ids >= shard_start) & (token_ids < shard_end)
local_idx = (token_ids - shard_start) * in_shard
local_val = local_logprobs.gather(-1, local_idx) * in_shard
token_logprobs = all_reduce_sum(local_val)          # [batch]
```

Wire bytes per shard, per step: `batch * 4`.

Requires a new `tenstorrent.sharded_index_gather` composite plus custom sharding rule that knows the shard offset and width along the gather axis. Highest novelty piece of compiler work here, since unlike topk there is no existing pattern for global index gather against a sharded source.

### 4. `count_ge` via shard local count plus `all_reduce`

```
local_count = (local_logprobs >= token_logprobs.unsqueeze(-1)).sum(-1)
token_ranks = all_reduce_sum(local_count)
```

Likely tractable without a new composite: `compare` is element wise and `sum` along the sharded vocab axis with `all_reduce` is a standard pattern. Worth a quick spike to confirm Shardy picks this lowering automatically before deciding whether a dedicated composite is needed.

## Validation plan

* Numerical parity against the current `all_gather` based path on the TP models used in [tt-xla#5004](https://github.com/tenstorrent/tt-xla/pull/5004) validation.
* End to end vLLM with logprobs enabled on 1D mesh (n300 dual chip) and 2D mesh (n300-llmbox, galaxy-wh-6u).
* Tracy perf capture comparing wire bytes per decode step with and without logprobs requested.
* Regression check: `logprobs=None` path unaffected on both mesh shapes.

## Dependencies

* [tt-mlir#8601](https://github.com/tenstorrent/tt-mlir/pull/8601) merged.
* [tt-xla#5004](https://github.com/tenstorrent/tt-xla/pull/5004) merged.

## Out of scope

* The non logprobs requested decode path, already at zero `all_gather` after [tt-xla#5004](https://github.com/tenstorrent/tt-xla/pull/5004).
* `compute_logprobs` for prompt logprobs (`_get_prompt_logprobs_dict`); behavior there should follow once the sampled logprobs path is migrated.
